### DISCUSSION TITLE
nameof(@object) == "@object" while [CallerArgumentExpression] injects "object"

### DISCUSSION THREAD

**Discussion by r-pankevicius** [dotnet/csharplang#7224] [General]
Hi,

I'd like to believe it's .net runtime issue but would like to check here first, maybe it's language specific.

Here is how I found this subtle difference: https://github.com/r-pankevicius/EnsureArg6/commit/9fc6856c9a69d6ef2b2382dae20537d3d3e89947

/r

PS: The method under test (with CallerArgumentExpressionAttribute) is:
https://github.com/r-pankevicius/EnsureArg6/blob/main/src/EnsureArg6/EnsureArg6.Types.cs#L92

**Comment by CyrusNajmabadi** [comment]
Can you provide a clearer repro that includes all referencedd symbols.  I have no idea where `[CallerArgumentExpression]` is even used in the code you linked to.

**Reply by r-pankevicius** [reply]
Look below: https://github.com/dotnet/csharplang/discussions/7224#discussioncomment-5994877

**Reply by r-pankevicius** [reply]
Oh, Okay, I see, I updated the topic:
https://github.com/r-pankevicius/EnsureArg6/blob/main/src/EnsureArg6/EnsureArg6.Types.cs#L92

**Comment by HaloFour** [comment]
This?

```cs
using System;
using System.Runtime.CompilerServices;

object @object = "foo";
M(nameof(@object));

void M(string x, [CallerArgumentExpression("x")] string? y = null) {
    Console.WriteLine(x); // prints "object"
    Console.WriteLine(y); // prints "nameof(@object)"
}
```

**Reply by CyrusNajmabadi** [reply]
If so, that's by design.  `nameof` gives you the name of the thing (which is just `object`).  Whereas `CallerArgument` gives you the actual syntactic argument passed along.  in this case `nameof(@object)`.

**Reply by r-pankevicius** [reply]
I think it's opposite, look at examples from me and @HaloFour.

**Reply by CyrusNajmabadi** [reply]
In the examples you have posted, things work as I would expect them to.

**Comment by r-pankevicius** [comment]
@HaloFour Yes, it's my case:
```csharp
using System;

namespace CallerArgumentExpressionVsNameoff
{
	internal class Program
	{
		internal static void Main(string[] args)
		{
			object @object = "foo";
			M(nameof(@object));
		}

		private static void M(object x, [CallerArgumentExpression("x")] string? y = null)
		{
			Console.WriteLine(x); // prints "object"
			Console.WriteLine(y); // prints "nameof(@object)"
		}
	}
}
```
